fluentd/out_syslog.rb

```diff
@@ -33,7 +33,8 @@ def configure(conf)
       if not conf['remote_syslog']
         raise Fluent::ConfigError.new("remote syslog required")
       end
-      @socket = UDPSocket.new
+      addrfamily = IPAddr.new(IPSocket.getaddress(@remote_syslog)).ipv4? ? ::Socket::AF_INET : ::Socket::AF_INET6
+      @socket = UDPSocket.new(addrfamily)
```

Comment:
We also need this in out_syslog_buffered.rb?

Reply:
That's where I misunderstood... :)  TCPSocket and UDPSocket in ruby are not symmetric...
TCPSocket does not take the address family, but it supports both families.

Here are some quick tests.

[TCP tests]  
Please ignore the connect(2) errors in the following tests.  Both "127.0.0.1" and "::1" are equally handled with no IPv4/IPv6 settings.
For comparison, non-existing host fails with "Name or service not known".
More specifically, if the IPv6 style IP address is not allowed in the socket, "::1" case should fail with an error like "Address family for hostname not supported".
```
irb(main):006:0> TCPSocket.new("127.0.0.1", 80)
Traceback (most recent call last):
        6: from /usr/bin/irb:23:in `<main>'
        5: from /usr/bin/irb:23:in `load'
        4: from /usr/share/gems/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):6
        2: from (irb):6:in `new'
        1: from (irb):6:in `initialize'
Errno::ECONNREFUSED (Connection refused - connect(2) for "127.0.0.1" port 80)

irb(main):007:0> TCPSocket.new("::1", 80)
Traceback (most recent call last):
        7: from /usr/bin/irb:23:in `<main>'
        6: from /usr/bin/irb:23:in `load'
        5: from /usr/share/gems/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        4: from (irb):7
        3: from (irb):7:in `rescue in irb_binding'
        2: from (irb):7:in `new'
        1: from (irb):7:in `initialize'
Errno::ECONNREFUSED (Connection refused - connect(2) for "::1" port 80)

irb(main):008:0> TCPSocket.new("bogus", 80)
Traceback (most recent call last):
        7: from /usr/bin/irb:23:in `<main>'
        6: from /usr/bin/irb:23:in `load'
        5: from /usr/share/gems/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        4: from (irb):8
        3: from (irb):8:in `rescue in irb_binding'
        2: from (irb):8:in `new'
        1: from (irb):8:in `initialize'
SocketError (getaddrinfo: Name or service not known)
```
[UDP tests]
By default UDPSocket handles IPv4 address.  Sending to "::1" using the socket fails as follows.
```
irb(main):009:0> UDPSocket.new
=> #<UDPSocket:fd 7, AF_INET, 0.0.0.0, 0>
irb(main):010:0> socket = UDPSocket.new
=> #<UDPSocket:fd 8, AF_INET, 0.0.0.0, 0>

irb(main):011:0> socket.send("abc", 0, "::1", 80)
Traceback (most recent call last):
        5: from /usr/bin/irb:23:in `<main>'
        4: from /usr/bin/irb:23:in `load'
        3: from /usr/share/gems/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):11
        1: from (irb):11:in `send'
SocketError (getaddrinfo: Address family for hostname not supported)
```
If we create AF_INET6 socket, then it works just fine.
```
irb(main):012:0> socket = UDPSocket.new(::Socket::AF_INET6)
=> #<UDPSocket:fd 9, AF_INET6, ::, 0>

irb(main):013:0> socket.send("abc", 0, "::1", 80)
=> 3
```